File: solutions/350.intersection-of-two-arrays-ii/intersection-of-two-arrays-ii.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
// functional way
// const intersect = (nums1, nums2) => {
//   return nums1.reduce(
//     (acc, e) => {
//       if (acc.nums2.includes(e)) {
//         acc.output.push(e)
//         delete acc.nums2[acc.nums2.indexOf(e)]
//       }
//       return acc
//     },
//     { nums2: nums2, output: [] }
//   ).output
// }

// cleaner way
const intersect = (nums1, nums2) => {
  //create table 
  let table = nums1.reduce((acc, e)=> {
    acc[e] = acc[e]
      ? acc[e]+1
      : 1
    return acc
  }, {})

  // filter
  return nums2.filter(e=>{
    if (table[e]){
      table[e]-=1
      return true
    }
    return false
  })
}

File: solutions/350.intersection-of-two-arrays-ii/intersection-of-two-arrays-ii.py
class Solution:
    def intersect(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        123
        34
        """
        m1 = {}
        m2 = {}
        ret = []

        for n1 in nums1:
            m1[n1] = m1.get(n1, 0) + 1

        for n2 in nums2:
            m2[n2] = m2.get(n2, 0) + 1

        for n in set(nums2):
            ret += [n] * min(m1.get(n, 0), m2.get(n, 0))

        return ret
